Add indexing pressure documentation (#59456)
elastic/elasticsearch: docs/reference/index-modules/indexing-pressure.asciidoc

[[index-modules-indexing-pressure]]
== Indexing Pressure

Indexing documents into Elasticsearch introduces system load in the form of
memory and CPU load. Each indexing operation includes coordinating, primary, and
replica stages. These stages can be performed across multiple nodes in the
cluster. If too much indexing work is introduced into the system, the cluster

> can become saturated.

Reviewer: maybe add that it would adversely affect other parts of the system (e.g. searches, cluster coordination, ...)

Indexing pressure is primarily generated by external operations such as indexing

> requests or CCR follow tasks. However, there is some internal indexing pressure

Reviewer: or internally by mechanisms such recoveries and cross-cluster replication.

> during a shard recovery or primary failover.

Reviewer: I would leave out this sentence (and mention recoveries  in the above one)

Elasticsearch internally monitors indexing load. When the load exceeds
certain limits, new indexing work will be rejected.

[float]

> === Indexing Stages

Reviewer: This overlaps somewhat with the `Basic write model` section of `data-replication.asciidoc` ( which is mostly missing the coordinating bits). I wonder if we can extend that one and link to it here?

> === Indexing Stages

Reviewer: The stages are sequential (first coordination, then primary, then replica).

What's missing in this doc is how the accounting works, i.e. that coordination accounting is only done once all subsequent stages are done, same for primary stage. This can be explained that ES internally retries operations and therefore keeps the operation in-memory until all subsequent stages have successfully completed.

External indexing operations go through three stages. The node receiving the
indexing request is the coordinating node. In the coordinating stage the node
performs any configured ingest pipelines, separates the request into individual
shard requests, and dispatches the requests to the primary shards.

In the primary stage, the primary shard node ensures that it is still the primary
(otherwise it reroutes the request to the new primary), waits on active shards,

> indexes the documents into its <<index-modules-translog,Translog>> and Lucene, and

Reviewer: into Lucene and the translog (switch order)

> dispatches to the replicas.

Reviewer: and replicates the writes to the replicas

Finally, in the replica stage, every available replica node indexes the documents

> into its <<index-modules-translog,Translog>> and Lucene.

Reviewer: switch Lucene and translog

[float]
=== Memory Limits

Elasticsearch exposes a node setting `indexing_pressure.memory.limit` which
restricts the number of bytes for outstanding indexing requests. Be default,

> this setting is configured to be 10% of the heap.

Reviewer: I wonder if we should add here that this setting is already generously sized, and that increasing it too much will have adverse effects on the rest of the system?

can become saturated. This can adversely impact other components of the system
such as searches, cluster coordination, background processing, etc.

Indexing pressure is primarily generated by external operations such as indexing
requests or internally by mechanisms such recoveries and cross-cluster
replication.

Elasticsearch internally monitors indexing load. When the load exceeds
certain limits, new indexing work will be rejected.

[float]
=== Indexing Stages

External indexing operations go through three stages: coordinating, primary, and
replication. This write model is explain <<basic-write-model,here>>.

[float]
=== Memory Limits

Elasticsearch exposes a node setting `indexing_pressure.memory.limit` which
restricts the number of bytes for outstanding indexing requests. Be default,
this setting is configured to be 10% of the heap.

At the beginning of each <<indexing stage,here>>, Elasticsearch accounts for the
bytes consumed by an indexing request. This accounting is only released at the
end of the indexing stage. This means that upstream stages will account for the
request overheard until all downstream stages are complete. For example, the

> coordinating request will remain accounted for until primary and replication

Reviewer: I think that talking about "replication stage" is confusing (instead of replica stage). For me, the lifetime of replication starts when a primary completes the local indexing work, and completes when the primary has received an answer from all replicas, certainly not what's meant here.

Note that replication stage is used  confusingly throughout this doc

stages are complete. The primary request will remain accounted for until the

> replication stage is complete.

Reviewer: mention here as well that the reason for this is to power internal retries.

[[index-modules-indexing-pressure]]
== Indexing pressure

> Indexing documents into Elasticsearch introduces system load in the form of
> memory and CPU load. Each indexing operation includes coordinating, primary, and
> replica stages. These stages can be performed across multiple nodes in the
> cluster. If too much indexing work is introduced into the system, the cluster
> can become saturated. This can adversely impact other components of the system
> such as searches, cluster coordination, background processing, etc.
> 
> Indexing pressure is primarily generated by external operations such as indexing
> requests or internally by mechanisms such recoveries and cross-cluster
> replication.
> 
> Elasticsearch internally monitors indexing load. When the load exceeds
> certain limits, new indexing work will be rejected.
> 

Reviewer 2: I re-ordered this to try to improve the flow.
```suggestion
Indexing documents into {es} introduces system load in the form of memory and
CPU load. Each indexing operation includes coordinating, primary, and replica
stages. These stages can be performed across multiple nodes in a cluster.

Indexing pressure can build up through external operations, such as indexing
requests, or internal mechanisms, such as recoveries and {ccr}. If too much
indexing work is introduced into the system, the cluster can become saturated.
This can adversely impact other operations, such as search, cluster
coordination, and background processing.

To prevent these issues, {es} internally monitors indexing load. When the load
exceeds certain limits, new indexing work is rejected.

```

> [float]
> === Indexing stages

Reviewer 2: ```suggestion
[discrete]
[[indexing-stages]]
=== Indexing stages
```

External indexing operations go through three stages: coordinating, primary, and

> replica. This write model is explained <<basic-write-model,here>>.

Reviewer 2: ```suggestion
replica. See <<basic-write-model>>.
```

> [float]
> === Memory limits

Reviewer 2: ```suggestion
[discrete]
[[memory-limits]]
=== Memory limits
```

> Elasticsearch exposes a node setting `.memory.limitindexing_pressure` which
> restricts the number of bytes for outstanding indexing requests. By default,
> this setting is configured to be 10% of the heap.

Reviewer 2: ```suggestion
The `.memory.limitindexing_pressure` node setting
restricts the number of bytes available for outstanding indexing requests.
This setting defaults to 10% of the heap.
```

> At the beginning of each indexing stage, Elasticsearch accounts for the

Reviewer 2: ```suggestion
At the beginning of each indexing stage, {es} accounts for the
```

bytes consumed by an indexing request. This accounting is only released at the
end of the indexing stage. This means that upstream stages will account for the
request overheard until all downstream stages are complete. For example, the
coordinating request will remain accounted for until primary and replica
stages are complete. The primary request will remain accounted for until each
in-sync replica has responded to enable replica retries if necessary.

> A node will start rejecting new indexing work at the coordinating or primary
> stage when the number of outstanding coordinating, primary, and replica indexing
> bytes are greater than the configured limit.

Reviewer 2: ```suggestion
A node will start rejecting new indexing work at the coordinating or primary
stage when the number of outstanding coordinating, primary, and replica indexing
bytes exceeds the configured limit.
```

A node will start rejecting new indexing work at the replica stage when the

> number of outstanding replica indexing bytes are greater than 1.5x the

Reviewer 2: ```suggestion
number of outstanding replica indexing bytes exceeds 1.5x the
```

configured limit. This design means that as indexing pressure builds on nodes,
they will naturally stop accepting coordinating and primary work in favor of
outstanding replica work.

> The default limit `indexing_pressure.memory.limit` (10%) is generously sized and
> should only be modified after careful consideration. Only indexing requests
> contribute to this limit meaning that there is additional indexing overhead
> (buffers, listeners, etc) which also require heap space. Finally, other
> components of Elasticsearch also require memory. Configuring this limit to be
> too high can starve other system components of operating memory.

Reviewer 2: ```suggestion
The `indexing_pressure.memory.limit` setting's 10% default limit is generously sized. You
should only change it after careful consideration. Only indexing requests
contribute to this limit. This means there is additional indexing overhead
(buffers, listeners, etc) which also require heap space. Other
components of {es} also require memory. Setting this limit 
too high can deny operating memory to other operations and components.
```

> [float]
> === Monitoring

Reviewer 2: ```suggestion
[discrete]
[[indexing-pressure-monitoring]]
=== Monitoring
```

> Indexing pressure metrics are exposed by the
> <<cluster-nodes-stats-api-response-body-indexing-pressure,Node Stats API>>.

Reviewer 2: ```suggestion
You can use the <<cluster-nodes-stats-api-response-body-indexing-pressure,node stats API>> to retrieve indexing pressure metrics.

```

> [float]
> === Indexing pressure settings

Reviewer 2: ```suggestion
[discrete]
[[indexing-pressure-settings]]
=== Indexing pressure settings
```

`indexing_pressure`::

>   A configurable limit for the number of outstanding bytes consumed by indexing
>   requests. New coordinating and primary operations will start being rejected
>   once this limit is hit. New replica operations will start being rejected when
>   replica operations consumed 1.5X this limit. Defaults to `10%` of the heap.

Reviewer 2: ```suggestion
(integer)
 Number of outstanding bytes that may be consumed by indexing requests. When
 this limit is reached or exceeded, the node will reject new coordinating and
 primary operations. When replica operations consume 1.5x this limit, the node
 will reject new replica operations. Defaults to 10% of the heap.
```

[[index-modules-indexing-pressure]]
== Indexing pressure

Indexing documents into Elasticsearch introduces system load in the form of
memory and CPU load. Each indexing operation includes coordinating, primary, and
replica stages. These stages can be performed across multiple nodes in the
cluster. If too much indexing work is introduced into the system, the cluster
can become saturated. This can adversely impact other components of the system
such as searches, cluster coordination, background processing, etc.

Indexing pressure is primarily generated by external operations such as indexing
requests or internally by mechanisms such recoveries and cross-cluster
replication.

Elasticsearch internally monitors indexing load. When the load exceeds
certain limits, new indexing work will be rejected.

[discrete]
[[indexing-stages]]
=== Indexing stages

External indexing operations go through three stages: coordinating, primary, and

> replica. See <<basic-write-model,here>>.

Reviewer: does this link need fixing?  (<<here>>)

Reviewer 2: I'd remove the `here` text:
```suggestion
replica. See <<basic-write-model>>.
```
